File: seleniumgrid_playwright/python/pw_cdp.py
import time

from playwright.sync_api import sync_playwright
from selenium import webdriver
from selenium.webdriver.chrome.options import Options

BASE = "localhost:4444"


def run_playwright_task(cdp_url: str, url_to_visit: str):
    with sync_playwright() as p:
        print(f"🎬 Opening {url_to_visit}")
        browser = p.chromium.connect_over_cdp(cdp_url)
        if len(browser.contexts) == 0:
            context = browser.new_context(no_viewport=True)
            page = context.new_page()
        else:
            context = browser.contexts[0]
            page = context.pages[0] if len(context.pages) > 0 else context.new_page()
        page.goto(url_to_visit, wait_until="networkidle")

        print(f"✅ Loaded: {page.title()}")

        time.sleep(2)

        return page.content()


class SeleniumGridBridge:
    def __init__(self):
        chrome_options = self._get_chrome_options()
        self.driver = webdriver.Remote(
            command_executor=f"http://{BASE}", options=chrome_options
        )

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.driver:
            print("Cleaning up: Closing Selenium driver...")
            self.driver.quit()

    def _get_chrome_options(self) -> Options:
        chrome_options = Options()
        chrome_options.add_experimental_option("excludeSwitches", ["enable-automation"])
        chrome_options.add_experimental_option("useAutomationExtension", False)
        chrome_options.headless = False
        arguments = [
            "--enable-features=NetworkService,NetworkServiceInProcess",
            "--no-sandbox",
            "--disable-blink-features=AutomationControlled",
            "--start-maximized",
            "--no-first-run",
        ]
        for argument in arguments:
            chrome_options.add_argument(argument)
        return chrome_options

    def get_cdp_url(self) -> str:
        return f"ws://{BASE}/session/{self.driver.session_id}/se/cdp"


if __name__ == "__main__":
    with SeleniumGridBridge() as bridge:
        url_to_visit = "https://www.google.com"
        cdp_url = bridge.get_cdp_url()
        content = run_playwright_task(cdp_url, url_to_visit)
        print(content)

File: seleniumgrid_playwright/js/pw_cdp.js
const { Builder } = require("selenium-webdriver");
const chrome = require("selenium-webdriver/chrome");
const { chromium } = require("playwright");

const BASE = "localhost:4444";

async function runPlaywrightTask(cdpUrl, urlToVisit) {
  const browser = await chromium.connectOverCDP(cdpUrl);
  const defaultContext = browser.contexts()[0];
  let page = defaultContext.pages()[0] || (await defaultContext.newPage());

  console.log(`🎬 Navigating to ${urlToVisit}...`);
  await page.goto(urlToVisit, { waitUntil: "networkidle" });

  const title = await page.title();
  console.log(`✅ Loaded: ${title}`);
  await new Promise((resolve) => setTimeout(resolve, 2000));
  const content = await page.content();
  return { content, playwrightBrowser: browser, defaultContext, page };
}

class SeleniumGridBridge {
  constructor() {
    this.driver = null;
  }

  async init() {
    const options = this._getChromeOptions();
    this.driver = await new Builder()
      .usingServer(`http://${BASE}`)
      .forBrowser("chrome")
      .setChromeOptions(options)
      .build();
    return this;
  }

  _getChromeOptions() {
    const options = new chrome.Options();
    options.excludeSwitches("enable-automation");
    options.setUserPreferences({ useAutomationExtension: false });
    options.addArguments(
      "--enable-features=NetworkService,NetworkServiceInProcess",
      "--no-sandbox",
      "--disable-blink-features=AutomationControlled",
      "--start-maximized",
      "--no-first-run",
    );
    return options;
  }

  async getCdpUrl() {
    const session = await this.driver.getSession();
    return `ws://${BASE}/session/${session.getId()}/se/cdp`;
  }

  async cleanup() {
    if (this.driver) {
      console.log("Cleaning up: Closing Selenium driver...");
      await this.driver.quit();
    }
  }
}

(async () => {
  const bridge = new SeleniumGridBridge();
  let pwBrowser = null;
  let pwContext = null;
  let pwPage = null;
  try {
    await bridge.init();
    const cdpUrl = await bridge.getCdpUrl();
    const result = await runPlaywrightTask(cdpUrl, "https://www.github.com");
    content = result.content; 
    pwBrowser = result.playwrightBrowser;
    pwContext = result.context;
    pwPage = result.page;

    console.log("📄 Content length:", content.length);
  } catch (error) {
    console.error("An error has ocurred:", error);
  } finally {
    if (pwPage) await pwPage.close();
    if (pwContext) await pwContext.close();
    if (pwBrowser) await pwBrowser.close();

    await bridge.cleanup();
  }
})();
